### PR TITLE
Debug/0 metric

### PR DESCRIPTION
**Pull request by pmhalvor**
No more zeros for RACL scores!

While the current setups aren't perfect, they all show steady improvements for both `absa`  and `hard` metrics.
This is an improvement on what I've been seeeing so far during experimenting. 

Now, some hyperparameter tuning should be done to actually get that metric a safe distance above 10%.

### CODE REVIEW

File: src/model.py [hunk 1]
```diff
@@ -807,21 +801,30 @@ def init_optimizer(self):
...
+                            optimizers[task].add_param_group(
+                                {"params": self.components["shared"][layer].parameters(), "lr":lr}
+                            )
+
+                    if self.find(f"bert_{task}", default=False) and self.finetune:
+                        optimizers[task].add_param_group(
+                            {"params": self.bert.parameters(), "lr":self.learning_rate}
+                        )
+
+                else:
```
Comment (pmhalvor) [inline src/model.py:823]: Not every task trains bert by default anymore. Now configurable via model param `bert_{task}`, as long as `bert_finetune=True` of course. 

NOTE: Should probably allow flexibility so that `bert_finetune` doesn't need to be called.

File: src/model.py [hunk 2]
```diff
@@ -1106,15 +1108,15 @@ def init_components(self, subtasks):
...
+                        out_features=2
                     ),
                     torch.nn.Sigmoid()
                 ).to(torch.device(self.device))
             })
         })
 
         self.relu = torch.nn.ReLU()
-        self.scope_loss = torch.nn.BCELoss()
+        self.scope_loss = torch.nn.CrossEntropyLoss(ignore_index=self.ignore_id)
```
Comment (pmhalvor) [inline src/model.py:1119]: Better loss for ignoring word pieces and paddings, preventing model confusing around what a scope is

File: src/model.py [hunk 3]
```diff
@@ -1250,15 +1251,18 @@ def forward(self, batch):
             queries, keys, values = self.get_attention_inputs(
                 target_cnn_output.permute(2, 0, 1),     # sequence, batch, embedding
                 expression_cnn_output.permute(2, 0, 1), 
-                polarity_cnn_output.permute(2, 0, 1)
+                polarity_cnn_output.permute(2, 0, 1),
+                batch
```
Comment (pmhalvor) [inline src/model.py:1255]: Now needs batch for golden transmission

File: src/model.py [hunk 4]
```diff
@@ -1284,9 +1288,10 @@ def forward(self, batch):
 
         return self.output
 
-
-    def get_attention_inputs(self, target, expression, polarity):
+    def get_attention_inputs(self, target, expression, polarity, batch):
         """ """
+        gold_transmission = self.find("gold_transmission", default=False)
+
```
Comment (pmhalvor) [inline src/model.py:1294]: Similar to the warm_up from IMN

File: src/model.py [hunk 5]
```diff
@@ -1316,16 +1321,29 @@ def get_attention_inputs(self, target, expression, polarity):
...
+        # BUG: size problem
+        # values is of size [batch, cnn_dim, sequence]
+        # true_values is of size [batch, sequence]
+        if gold_transmission:
+            gold_influence = self.get_prob(self.current_epoch, self.find("warm_up_constant", default=5))
+            values = (  # strengthen signal for true values
+                gold_influence*true_values.bool().float().unsqueeze(-1).expand(values.shape)
+                 + (1-gold_influence)*values
+            ).detach().to(torch.device(self.device))
+            values.requires_grad = True
```
Comment (pmhalvor) [inline src/model.py:1343]: The tokens where a label should be present are extra loud in attention at beginning, to guide model in correct direction.

File: src/model.py [hunk 6]
```diff
@@ -1342,26 +1360,31 @@ def scope_relevance(self, batch, shared_output) -> tuple():
...
         self.scope_loss_value.backward(retain_graph=True)
         self.scope_optimizer.step()
 
         # TODO configurable guided start/warm_up?
         gold_influence = self.get_prob(self.current_epoch, self.find("warm_up_constant", default=5))
-        self.scope_output = (gold_influence*self.scope_true + (1-gold_influence)*self.scope_logits).detach()
+        try:  # FIXME remove this try/catch
+            self.scope_output = (
+                gold_influence*self.scope_true + (1-gold_influence)*self.scope_logits.argmax(-1).squeeze(-1)
+            ).detach()
```
Comment (pmhalvor) [inline src/model.py:1379]: Scope also recieves a golden transmission, to guide model into predicting correct scope at begging of training

File: src/model.py [hunk 7]
```diff
@@ -1377,21 +1400,100 @@ def get_prob(epoch_count, constant=5):
...
+                "target_expression": torch.nn.MultiheadAttention(
+                    embed_dim = cnn_dim,
+                    num_heads = 1,
+                    dropout=self.dropout,
+                    # query: torch.norm(target)
+                    # key: torch.norm(expression)
+                    # values: target
+                    # key_padding_mask: (batch[1]*-1)+1
+
+                    # qk-outputs used in group v-mul later
```
Comment (pmhalvor) [inline src/model.py:1419]: Pick up here, implementing the different relation-type in RACL

**Review by pmhalvor — COMMENTED** [review]
Working, ready for merge. See comments for major changes.
